swap/src/alice/swap.rs

```diff
@@ -85,267 +86,334 @@ pub enum AliceState {
     SafelyAborted,
 }
 
+impl fmt::Display for AliceState {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AliceState::Started { .. } => write!(f, "started"),
+            AliceState::Negotiated { .. } => write!(f, "negotiated"),
+            AliceState::BtcLocked { .. } => write!(f, "btc_locked"),
+            AliceState::XmrLocked { .. } => write!(f, "xmr_locked"),
+            AliceState::EncSignLearned { .. } => write!(f, "encsig_sent"),
+            AliceState::BtcRedeemed => write!(f, "btc_redeemed"),
+            AliceState::BtcCancelled { .. } => write!(f, "btc_cancelled"),
+            AliceState::BtcRefunded { .. } => write!(f, "btc_refunded"),
+            AliceState::Punished => write!(f, "punished"),
+            AliceState::SafelyAborted => write!(f, "safely_aborted"),
+            AliceState::BtcPunishable { .. } => write!(f, "btc_punishable"),
+            AliceState::XmrRefunded => write!(f, "xmr_refunded"),
+            AliceState::WaitingToCancel { .. } => write!(f, "waiting_to_cancel"),
+        }
+    }
+}
+
+pub async fn swap(
+    state: AliceState,
+    swarm: Swarm,
+    bitcoin_wallet: Arc<crate::bitcoin::Wallet>,
+    monero_wallet: Arc<crate::monero::Wallet>,
+) -> Result<(AliceState, Swarm)> {
+    run_until(state, is_complete, swarm, bitcoin_wallet, monero_wallet).await
+}
+
+// TODO: use macro or generics
+pub fn is_complete(state: &AliceState) -> bool {
+    matches!(
+        state,
+        AliceState::XmrRefunded
+            | AliceState::BtcRedeemed
+            | AliceState::Punished
+            | AliceState::SafelyAborted
+    )
+}
+
+pub fn is_xmr_locked(state: &AliceState) -> bool {
+    matches!(
+        state,
+        AliceState::XmrLocked{..}
+    )
+}
+
 // State machine driver for swap execution
 #[async_recursion]
-pub async fn swap(
+pub async fn run_until(
     state: AliceState,
+    is_state: fn(&AliceState) -> bool,
```

Comment:
```suggestion
    is_target_state: fn(&AliceState) -> bool,
```
`is_state` is a bit confusing per se.